chapters/chapter_5/section_5.md

```diff
@@ -172,233 +318,157 @@ dummy_worker_insert(LogThreadedDestDriver *d, LogMessage *msg)
-
-  self->super.worker.thread_init = dummy_worker_thread_init;
-  self->super.worker.thread_deinit = dummy_worker_thread_deinit;
-  self->super.worker.disconnect = dummy_dd_disconnect;
-  self->super.worker.insert = dummy_worker_insert;
-
-  self->super.format_stats_instance = dummy_dd_format_stats_instance;
-  self->super.super.super.super.generate_persist_name = dummy_dd_format_persist_name;
-  //self->super.stats_source = SCS_DUMMY;
+  /*
+    If you created resources during new,
+    you need to free them here.
+  */
 
-  return (LogDriver *)self;
+  log_threaded_dest_worker_free_method(s);
 }
 
-extern CfgParser dummy_parser;
-
-static Plugin dummy_plugin =
+LogThreadedDestWorker *
+example_destination_dw_new(LogThreadedDestDriver *o, gint worker_index)
 {
-  .type = LL_CONTEXT_DESTINATION,
-  .name = "dummy",
-  .parser = &dummy_parser,
-};
+  ExampleDestinationWorker *self = g_new0(ExampleDestinationWorker, 1);
 
-gboolean
-dummy_module_init(PluginContext *context, CfgArgs *args)
-{
-  plugin_register(context, &dummy_plugin, 1);
+  log_threaded_dest_worker_init_instance(&self->super, o, worker_index);
+  self->super.thread_init = _thread_init;
+  self->super.thread_deinit = _thread_deinit;
+  self->super.insert = _dw_insert;
+  self->super.free_fn = _dw_free;
+  self->super.connect = _connect;
+  self->super.disconnect = _disconnect;
 
-  return TRUE;
+  return &self->super;
 }
-
-const ModuleInfo module_info =
-{
-  .canonical_name = "dummy",
-  .version = SYSLOG_NG_VERSION,
-  .description = "This is a dummy destination for syslog-ng.",
-  .core_revision = SYSLOG_NG_SOURCE_REVISION,
-  .plugins = &dummy_plugin,
-  .plugins_len = 1,
-};
 ```
 
-This unit can be separated into 5 parts:
-
-1. configuration functions (setters)
-2. utility functions
-3. functions running in separate thread
-4. functions running in the main thread
-5. plugin-construction declarations
-
-Functions from category 2-4 are implementations of the LogThreadedDestDriver's virtual methods.
-In `dummy_dd_new`, we pass these function pointers to the base classes.
-
-Our example overrides these 6 virtual methods:
+Our example overrides these virtual methods:
+- `thread_init (_thread_init)`: if you need to initialize thread specific resources, you can do them here. We are saving the thread id in this example. Non-thread specific resources may be created in the constructor (`example_destination_dw_new`).
+- `thread_deinit (_thread_deinit)`: if you created resources during `thread_init`, you need to deallocate them here.
+- `free_fn (_dw_free)` destructor for resources created in `example_destination_dw_new`.
+- `insert (_dw_insert)`: It formats the received message and sends to an actual destination.
+- `connect (_dw_connect)`: This is called after `thread_init`, before the first `insert`, and each time you signal broken connection (returning `LTR_NOT_CONNECTED` from `insert`). You can open files or establish connections here.
```

Comment:
LTR_ERROR and LTR_DROP causes `connect` to be called too

Reply:
Fixed thanks